### PR TITLE
Feat/test kitchen ci

### PR DESCRIPTION
**Pull request by Lirt**
Hello,

I created test-kitchen configuration file for Vagrant driver.

Test Kitchen is used to automatize integration tests of configuration management tools. In this case I set up `.kitchen.yml` fileand there are 2 tests. First one is creation of tor non-exit node and second creation of tor exit node. Configurations of tests are located in `test/integration/default/`.

With Test-Kitchen you can automatically run deployment test of tor relay on almost all systems that you officialy support with exception of OpenBSD which is not supported by kitchen-vagrant driver (Bento repository). Vagrant tests are close to real world use case and they can be done only from your local machine (not using Travis or other CI). Just follow `Testing` part of `README.md` for instruction to use it. 

Notice that during tests I use `tor_DisableNetwork: 1` and `tor_PublishServerDescriptor: 0` to do dry run of tests.

In future we can also test multiple versions of ansible.

This PR was created during my hackaton partly for educational purposes, but I am using your repository and plan to use it more widely during automated deployment of Tor nodes.

Feel free to ask me anything during review.

Kindest Regards
Ondrej

### CODE REVIEW

**Comment by nusenu** [comment]
Hi Ondrej,

thanks for your big PR.
Travis-ci's requested permissions got a lot more reasonable since I last checked, great.

Regarding: https://github.com/nusenu/ansible-relayor/pull/149/commits/dfbb366d836007fa00a87d6c41c3b3f60eed1838 
we usually do not reiterate tor's man page.  It will cause a redundant torrc change (add a new line to torrc files because that var is now defined but was not previously) for all deployments and a reload. What is your motivation for this commit? Is it required?

 * it would be nice to have a rebase shortly before I have more time to look into it
 * please squash multiple commits on a single new file into a single commit

**Comment by Lirt** [comment]
Hi nusenu,

I can checkout back dfbb366 or use official tor manual page text there. Motivation was to document my usage of this option but I agree it should copy manpage text.

I will look into squashing shortly.

**Comment by nusenu** [comment]
> I can checkout back dfbb366 

Yes please remove it (since we do not copy all the other torrc manual entries 
into defaults/main.yml either). The tor manual page is the authoritative source for that.
Removing it will also avoid the reload.


some more thoughts:

 * consider adding PublishServerDescriptor 0 in addition to DisableNetwork
 * lets start with vagrant only and split docker into a separate PR that would
be on top of this one but I'm not entirely convinced about adding kitchen-docker
due to its limitations (systemd, FreeBSD).

just putting this reference here
https://stackoverflow.com/questions/42852457/kitchen-docker-and-systemd#comment72811869_42852457


thanks for your work so far!

**Comment by Lirt** [comment]
Changes:

* Removed commit with torrc changes for `tor_DisableNetwork: 1`
* Added `tor_PublishServerDescriptor: 0`
* Rebased whole PR to address only Test Kitchen tests with Vagrant driver.

This is now clean PR with no changes in Ansible code. 

I will try to make Docker + Travis CI PR cleaner in new PR as soon as I find solution.

**Comment by nusenu** [comment]
Please squash these commits:

1) .kitchen.yml
0778013cff8aa0fc3dca590641bd2aa05db157f1
59d627903b4fd32fa766d824e571dbe2faf9d03a
e3a6b8f1bf22b3e22ac3f94e0bfd19e5f16a2e09
f2d46505410f4f830637c8ff236a6db02bbd8c1d

2) test/integration/default/default.yml
a6907d4e58ac8caa45842c7247b9831f9a032f1f
c83580aecf16ca05c1108f84e0168dfcfe0d003c
649d7d6337b22fa655c61e78054c253c2cfd1484

**Comment by Lirt** [comment]
My first squashing, hope this is fine. 

Btw. for Docker + Travis CI, I think I will abandon this, because of systemd issues, where I can't even make it work the hacky way.

**Comment by nusenu** [comment]
> My first squashing, hope this is fine.

Looks good.

since the files match:
test/vagrant_provisions/freebsd10_vagrant_provision.rb
test/vagrant_provisions/freebsd11_vagrant_provision.rb

lets replace it with a single 

test/vagrant_provisions/freebsd_vagrant_provision.rb



Lets link the file instead of copying it:
cp /usr/local/bin/python /usr/bin/python

**Comment by nusenu** [comment]
* `kitchen-vagrant` is missing in the `gem install` line (README)
* lets remove non-default lines from the default suite:
```
    tor_signingkeylifetime_days: 90
    tor_nickname: "{{ ansible_hostname }}"
    tor_ports:
        - { orport: 443, dirport: 80}
```
* Please add OpenBSD to the list of platforms + python installation and link
```
- name: openbsd-6.2
  driver:
    box: generic/openbsd6
```

* there is an exit-node.yml but it is not added in the suits section of .kitchen.yml?
* maybe reduce the memory size per VM to 500MB?
* can we limit the amount of concurrently running VMs (to let tests run sequentially if people do not have enough RAM for all VMs)? 

* rebase to the latest master and squash all commits into a single commit with a single clean commit message, then we are probably ready

thank you!

**Comment by nusenu** [comment]
I was not able to find the source of the "generic" vagrant cloud user, so lets skip OpenBSD for now.
https://groups.google.com/d/msg/vagrant-up/99OzkAJ5OgA/UWDpTZDhAQAJ

**Comment by Lirt** [comment]
Thanks for notes, I will look into that tomorrow. 

Vagrant driver for Kitchen is by default using images from [Bento](https://app.vagrantup.com/bento) so I picked images that you support and Bento delivers. However I can try to make the generic one work.

**Comment by Lirt** [comment]
Should I remove squashed commit messages into one message about all changes in this full commit?

**Comment by nusenu** [comment]
* yes, please replace the commit message with a single description about what this commit is about and what platforms it covers
* lets add `openbsd_vagrant_provision.rb` only once we add openbsd to the platform list (maybe we can use https://app.vagrantup.com/trombik/boxes/ansible-openbsd-6.2-amd64)
* I don't know if that makes any difference but lets use ```Vagrant.configure("2")```  (with "") for consistency with other places (documentation)

If bento accepts my PR for HardenedBSD we can add that platform in the future
https://github.com/nusenu/bento/commit/f7b857a55102ecde57e657462c435a38a29b6f51

**Comment by nusenu** [comment]
We will have these lines in all ansible yml files, can we have them in one place instead? (I'll add a lot more test cases in the future and would like to avoid duplicate lines)
```
tor_ContactInfo: "ansible-relayor test-kitchen (you should never see this on a public relay) https://github.com/nusenu/ansible-relayor"
tor_DisableNetwork: 1
tor_PublishServerDescriptor: 0
```

**Comment by Lirt** [comment]
I can think about including playbook file with variables which you want to share. For example using [include_vars](http://docs.ansible.com/ansible/latest/include_vars_module.html)

**Comment by nusenu** [comment]
* something appears to be confused by the "-" in the second suite name "exit-node" -> lets rename the "exit-node" suite to "exit" otherwise ansible shows them as "node-debian-9" (instead of exit-node-debian-9)

**Comment by nusenu** [comment]
> something appears to be confused by the "-" in the second suite name "exit-node"
I'm wondering if this is a bug that should be reported upstream

File: test/integration/default/exit-node.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+- hosts: all
+  vars_files:
+    - vars/dry-run-vars.yml
+  vars:
+    tor_ExitRelay: "True"
```
Comment (nusenu) [inline test/integration/default/exit-node.yml]: it should be:
```tor_ExitRelay: True```
(without "")

**Comment by Lirt** [comment]
This looks like issue with how the `kitchen_ansiblepush` provisioner is generating hostnames of Vagrant instances to inventory. 

You can list ansible inventory of kitchen ansible push provisioner with script found by `whereis kitchen-ansible-inventory`.

I thought that this might have something with using dashes instead of underscores, but it does not. If I understand that correctly, this line means, that first word before dash is dropped https://github.com/ahelal/kitchen-ansiblepush/blob/master/lib/kitchen/provisioner/ansible_push.rb#L101

Sorry for the double-quotes, it was result of yamllinter.

**Comment by nusenu** [comment]
I made some modifications to the commit message and committed it,
thank you!

https://github.com/nusenu/ansible-relayor/commit/4f4e3bbe37126c680028142878fd1fa1f9ab706d

I'll come up with a naming scheme for suites, probably something like:
* t-non-exit
* t-exit
* t-non-exit-alpha
* t-exit-alpha
...

**Comment by nusenu** [comment]
This is now in master.
